File: mrvtools/mrvtools/report/adaptation_tracking_report/adaptation_tracking_report.py
# Copyright (c) 2023, tridotstech and contributors
# For license information, please see license.txt

import frappe


def execute(filters=None):
	columns, data, chart = get_columns(), get_datas(filters), get_chart()
	return columns, data, None, chart


def get_columns():
	col = []
	col.append("Project ID" + ":Link/Project")
	col.append("Project Title" + ":Data")
	col.append("Objective" + ":Data")
	col.append("Key Sector" + ":Data")
	col.append("Key Sub-sector" + ":Data")
	col.append("Cost in USD" + ":Float")
	col.append("Location" + ":Data")
	col.append("Implementing entity or entities" + ":Data")
	col.append("Other Agency" + ":Data")
	col.append("Start Date" + ":Data")
	col.append("Lifetime in Years" + ":Int")
	col.append("Included In" + ":Data")
	col.append("Impact Summaries" + ":Data")
	return col

def get_datas(filters):
	conditions = ""
	if filters.get("year"):
		frappe.log_error("Not Empty","Not Empty")
		conditions += f" AND  YEAR(A.financial_closure_date) <= '{filters.get('year')}'"
	if filters.get("year") ==None:
		frappe.log_error("Empty","Empty")
		conditions += f"AND A.financial_closure_date <= '{frappe.utils.today()}'"
	if filters.get("key_sector"):
		conditions += f" AND P.key_sector like '{filters.get('key_sector')}'"
	if filters.get("key_sub_sector"):
		conditions += f" AND P.key_sub_sector like '{filters.get('key_sub_sector')}'"
	if filters.get("impact_summaries"):
		cur_ele = filters.get("impact_summaries").lower()
		cur_word = cur_ele.split(" ")
		resWord = "_".join(cur_word)
		conditions += f" AND A.{resWord} = 1"


	query= f"""
			SELECT
				P.name as project_id,
				P.project_name as project_title,
				P.objective,
				P.key_sector,
				P.key_sub_sector,
				P.costusd as cost_in_usd,
				P.location,
				P.implementing_entity as implementing_entity_or_entities,
				P.other_agency,
				P.start_date,
				P.lifetime as lifetime_in_years,
				A.included_in,
				A.name
				
			FROM
				`tabProject` as P
			INNER JOIN
				`tabAdaptation` as A 
			ON
				P.name = A.project_name
			WHERE 
				P.docstatus != 2
				{conditions}
			ORDER BY
				A.project_name
	"""
	result = frappe.db.sql(query, as_dict=1)
	field_list = []
	meta = frappe.get_meta("Adaptation")
	meta_dict = meta.as_dict()
	fields = meta_dict["fields"]
	for field in fields:
		if field["fieldtype"] == "Check":
			field_list.append(field["fieldname"])
	for each in result:
		get_list=[]
		values= frappe.db.get_value("Adaptation",each["name"],field_list,as_dict=1)
		for key,value in values.items():
			if value == 1:
				cur_ele = key.split("_")
				cur_word = " ".join(cur_ele)
				resu = cur_word.title()
				get_list.append(resu)
		resString = ",".join(get_list)
		each["impact_summaries"] = resString
	return result

def get_chart():
	field_list = []
	get_counts = []
	meta = frappe.get_meta("Adaptation")
	meta_dict = meta.as_dict()
	fields = meta_dict["fields"]
	for field in fields:
		if field["fieldtype"] == "Check":
			field_list.append(field["fieldname"])
	for field in field_list:
		count=0
		values=frappe.db.get_all("Adaptation",field)
		for i in values:
			for key,value in i.items():
				if value ==1:
					count= count+1
		get_counts.append(count)
		
	chart = {
		'data': {
		  'labels': ['Forest Health', 'Watershed Health', 'Coastal Health', 'Water Security', 'Security of Place', 'Energy Security', 'Income Security', 'Community Health', 'Food Security'], 
		  'datasets': [{'title': 'No. of Projects', 'values': get_counts}]
		  }, 
		'type': 'bar', 
		'colors': ['red']
	}
	return chart


		

File: mrvtools/mrvtools/report/adaptation_tracking_report/adaptation_tracking_report.js
// Copyright (c) 2023, tridotstech and contributors
// For license information, please see license.txt
/* eslint-disable */
frappe.query_reports["Adaptation Tracking Report"] = {
	"filters": [
		{
			"fieldname": "year",
			"fieldtype": "Select",
			"label": __("Year"),
			"options": [" ","1990","1991","1992","1993","1994","1995","1996","1997","1998","1999","2000","2001","2002","2003","2004","2005","2006","2007","2008","2009","2010","2011","2012","2013","2014","2015","2016","2017","2018","2019","2020","2021","2022","2023","2024","2025","2026","2027","2028","2029","2030","2031","2032","2033","2034","2035","2036","2037","2038","2039","2040","2041","2042","2043","2044","2045","2046","2047","2048","2049","2050"]
		},
		{
			"fieldname": "key_sector",
			"fieldtype": "Link",
			"label": __("Key Sector"),
			"options":"Project Key Sector",
			get_query: function (doc) {
				return {
					"filters": {
						"objective": ["in", ['Cross-Cutting','Adaptation']]
					}
				}
			}
		},
		{
			"fieldname": "key_sub_sector",
			"fieldtype": "Link",
			"label": __("Key Sub-Sector"),
			"options":"Project Key Sub Sector",
			"depends_on":"eval:doc.key_sector",
			get_query: function () {
				return {
					"filters": {
						// "key_sector":["like",["%Cross-Cutting%","%Adaptation%"]]
						"key_sector": ["in", [`${frappe.query_report.filters[1].last_value}`]]
					}
					
				}
			}
		},
		{
			"fieldname": "impact_summaries",
			"fieldtype": "Select",
			"label": __("Impact Area"),
			"options":['','Forest Health', 'Watershed Health', 'Coastal Health', 'Water Security', 'Security of Place', 'Energy Security', 'Income Security', 'Community Health', 'Food Security']
		}
	]
};
